pallets/maintenance-mode/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+//!
+//! This introduces one storage read to fetch the base filter for each extrinsic. Hoever, it should
+//! be that the state cache eliminates this cost almost entirely. I wonder if that can or should be
+//! reflected in the weight calculation.
+//!
+//! This could be more configureable by letting the runtime developer specify a type (probably an
+//! enum) that can be converted into a filter and an origin.
+
+#![allow(non_camel_case_types)]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+use frame_support::{pallet, weights::Weight};
+
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_support::traits::{EnsureOrigin, Filter};
+	use frame_system::pallet_prelude::*;
+
+	/// Pallet for migrations
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// The base call filter to be used in normal operating mode
+		/// (When we aren't in the middle of a migration)
+		type NormalCallFilter: Filter<Self::Call>;
+		/// The base call filter to be used when we are in the middle of migrations
+		/// This should be very restrictive. Probably not allowing anything except possibly
+		/// something like sudo or other emergency processes
+		type MaintenanceCallFilter: Filter<Self::Call>;
+		/// The origin from which the call to enter or exit maintenance mode must come
+		/// Take care when choosing your maintenance call filter to ensure that you'll still be
+		/// able to return to normal mode. For example, if your MaintenanceOrigin is a council, make
+		/// sure that your councilors can still cast votes.
+		type MaintenanceOrigin: EnsureOrigin<Self::Origin>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// The chain was put into Maintenance Mode
+		EnteredMaintenanceMode,
+		/// The chain returned to its normal operating state
+		NormalOperationResumed,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn maintenance_mode)]
+	/// Whether the site is in maintenance mode
+	type MaintenanceMode<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn previous_timestamp)]
+	/// The timestamp recorded on the previous block. Used to automatically enter maintenance mode
+	/// after a long delay in block production
+	type PreviousTimestamp<T: Config> = StorageValue<_, u64, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			// Idea: Detect if too much time has elapsed since the last block, and enter maintenance
+			// mode automatically.
```

Comment:
This is a nice idea, but I feel like it could cause some unintended problems. I like the manual approach because we have complete control, it feels more predictable.

Reply:
I totally agree about unintended problems. I have that worry too. Like some stall happens, and we figure out how to fix it, but now we have a secondary problem that we're stuck in maintenance mode for some reason.

The one thing that makes me think maybe the time detector is worth keeping is that it allows us to enter maintenance mode immediately after the stall ends before _any_ transactions are executed. If for some reason there undesirable transactions in the pool from during the stall, they would be competing priority-wise with governance transactions that propose or vote for the proposal to enter maintenance mode.

Reply:
Makes sense

Reply:
@crystalin What do you think about this feature idea? I'm fine to proceed with or without it.

Reply:
I think the feature is good, but the risk is too high/unpredictable.
We can probably afford to have emergency proposal with high priority when setting it to compensate

Reply:
If this is a Unix timestamp, I also wonder if it could be abused by a malicious collator.

Reply:
Sounds good. I agree. I'll remove it from the pallet.